Objects/longobject.c

```diff
@@ -1758,6 +1759,22 @@ long_to_decimal_string_internal(PyObject *aa,
     size_a = Py_ABS(Py_SIZE(a));
     negative = Py_SIZE(a) < 0;
 
+    /* quick and dirty pre-check for overflowing the digit limit,
+       based on the inequality 10/3 >= log2(10) */
+    if (size_a >= 10 * _PY_LONG_MAX_STR_DIGITS_THRESHOLD
+                  / (3 * PyLong_SHIFT) + 2) {
+        PyInterpreterState *interp = _PyInterpreterState_GET();
+        int max_str_digits = interp->int_max_str_digits;
+        if ((max_str_digits > 0) &&
+            /* avoid overflow in 10 * max_str_digits */
+            (max_str_digits <= INT_MAX / 10) &&
```

Comment:
This isn't ideal - it means that for huge `max_str_digits`, the limit gets ignored entirely. I think by being a bit cruder in the check we can rework this and still avoid integer overflow. I'll do that.